### PR TITLE
[HttpClient] Fix encoding some characters in query strings

### LINKED ISSUE #50670: HttpClient parseUrl is broken and not RFC compilant
### Symfony version(s) affected

5.4.24

### Description

When I use HttpClient with url which contain this data a part of path segment:
```
custom|2010-01-01 00:00:00|2023-06-15 05:50:35
```

I have got error from calling part because of wrongly url encoded data.

Issue is caused by this pull https://github.com/symfony/symfony/pull/49299

### How to reproduce

Calling  `HttpClient::parseUrl('http://some.url/custom|2010-01-01 00:00:00|2023-06-15 05:50:35?a=b#c)'`  should return result array:
```
Array
(
    ...
    [path] => /custom%7C2010-01-01%2000:00:00%7C2023-06-15%2005:50:35
    ...
)
```

### Possible Solution

_No response_

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by Kozzi11**
| Q             | A
| ------------- | ---
| Branch?       | 5.4 
| Bug fix?      | yes
| New feature?  | no 
| Deprecations? | no 
| Tickets       | Fix #50670 
| License       | MIT
| Doc PR        | symfony/symfony-docs#...

### CODE REVIEW

File: src/Symfony/Component/HttpClient/HttpClientTrait.php [hunk 1]
```diff
@@ -625,20 +625,18 @@ private static function mergeQueryString(?string $queryString, array $queryArray
                 // https://tools.ietf.org/html/rfc3986#section-2.3 + some chars not encoded by browsers
                 $queryString = strtr($queryString, [
                     '%21' => '!',
+                    '%23' => '#',
```
Comment (stof) [inline src/Symfony/Component/HttpClient/HttpClientTrait.php]: this is wrong. You cannot revert the encoding of `#` in the query string as the `#` would then mark the end of the query string (starting the fragment part)

File: src/Symfony/Component/HttpClient/HttpClientTrait.php [hunk 2]
```diff
@@ -625,20 +625,18 @@ private static function mergeQueryString(?string $queryString, array $queryArray
...
                     '%24' => '$',
                     '%28' => '(',
                     '%29' => ')',
                     '%2A' => '*',
                     '%2F' => '/',
                     '%3A' => ':',
                     '%3B' => ';',
+                    '%3F' => '?',
```
Comment (stof) [inline src/Symfony/Component/HttpClient/HttpClientTrait.php]: I'm quite sure browsers keep `?` escaped in the query string to avoid risk of bad parsing.

**Review by stof — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/HttpClient/HttpClientTrait.php (hunk 2 above)
Comment (Kozzi11) [reply src/Symfony/Component/HttpClient/HttpClientTrait.php]: good catch make sense I have remove this change

File: src/Symfony/Component/HttpClient/HttpClientTrait.php (hunk 1 above)
Comment (Kozzi11) [reply src/Symfony/Component/HttpClient/HttpClientTrait.php]: Yes you are right I have removed that

File: src/Symfony/Component/HttpClient/HttpClientTrait.php [hunk 3]
```diff
@@ -547,7 +547,7 @@ private static function parseUrl(string $url, array $query = [], array $allowedS
             }
 
             // https://tools.ietf.org/html/rfc3986#section-3.3
-            $parts[$part] = preg_replace_callback("#[^-A-Za-z0-9._~!$&/'()[\]*+,;=:@\\\\^`{|}%]++#", function ($m) { return rawurlencode($m[0]); }, $parts[$part]);
+            $parts[$part] = preg_replace_callback("#[^-A-Za-z0-9._~!$&/'()[\]*+,;=:@?\#%]++#", function ($m) { return rawurlencode($m[0]); }, $parts[$part]);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpClient/HttpClientTrait.php]: This force-encodes characters that are not in the list.
Which means with this change, `?` and `#` won't be encoded. This looks incorrect to me (even if I don't expect these to be found in parts of valid URLs).
Also *when the original URL contain these unencoded*, there is no need to force-encode the chars removed in the diff.
Which means I would keep the current regexp.

File: src/Symfony/Component/HttpClient/Tests/HttpClientTraitTest.php [hunk 4]
```diff
@@ -70,8 +70,8 @@ public static function provideResolveUrl(): array
             [self::RFC3986_BASE, '/g',            'http://a/g'],
             [self::RFC3986_BASE, '//g',           'http://g/'],
             [self::RFC3986_BASE, '?y',            'http://a/b/c/d;p?y'],
-            [self::RFC3986_BASE, '?y={"f":1}',    'http://a/b/c/d;p?y={%22f%22:1}'],
-            [self::RFC3986_BASE, 'g{oof}y',       'http://a/b/c/g{oof}y'],
+            [self::RFC3986_BASE, '?y={"f":1}',    'http://a/b/c/d;p?y=%7B%22f%22:1%7D'],
+            [self::RFC3986_BASE, 'g{oof}y',       'http://a/b/c/g%7Boof%7Dy'],
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpClient/Tests/HttpClientTraitTest.php]: these changes should be reverted, per my comment above

**Review by nicolas-grekas — COMMENTED** [review]
Unfortunately, the RFC is very loose on the topic, and implementations varies.

File: src/Symfony/Component/HttpClient/HttpClientTrait.php (hunk 3 above)
Comment (Kozzi11) [reply src/Symfony/Component/HttpClient/HttpClientTrait.php]: Ok can we remove `|` char from original regexp, this one is the problematic one, which cause issues for us.

Comment (Kozzi11) [reply src/Symfony/Component/HttpClient/HttpClientTrait.php]: And I still do not understand why the original PR https://github.com/symfony/symfony/pull/49299 has been merged to 5.4 branch? Maybe rever it from 5.4 branch would be ok too?

**Comment by nicolas-grekas** [comment]
Thank you @Kozzi11.

**Pull request merged**